util/rowcodec/encoder.go

```diff
@@ -0,0 +1,314 @@
+type Encoder struct {
+	row
+	tempColIDs []int64
+	values     []types.Datum
+	tempData   []byte
+	sc         *stmtctx.StatementContext
+}
+
+func (encoder *Encoder) reset() {
+	encoder.large = false
+	encoder.numNotNullCols = 0
+	encoder.numNullCols = 0
+	encoder.data = encoder.data[:0]
+	encoder.tempColIDs = encoder.tempColIDs[:0]
+	encoder.values = encoder.values[:0]
+}
+
+func (encoder *Encoder) addColumn(colID int64, d types.Datum) {
+	if colID > 255 {
+		encoder.large = true
+	}
+	if d.IsNull() {
+		encoder.numNullCols++
+	} else {
+		encoder.numNotNullCols++
+	}
+	encoder.tempColIDs = append(encoder.tempColIDs, colID)
+	encoder.values = append(encoder.values, d)
+}
+
+// Encode encodes a row from a datums slice.
+func (encoder *Encoder) Encode(colIDs []int64, values []types.Datum, buf []byte) ([]byte, error) {
+	encoder.reset()
+	for i, colID := range colIDs {
+		encoder.addColumn(colID, values[i])
+	}
+	return encoder.build(buf[:0])
+}
+
+func (encoder *Encoder) build(buf []byte) ([]byte, error) {
+	r := &encoder.row
+	// Separate null and not-null column IDs.
+	numCols := len(encoder.tempColIDs)
+	nullIdx := numCols - int(r.numNullCols)
+	notNullIdx := 0
+	if r.large {
+		encoder.initColIDs32()
+		encoder.initOffsets32()
+	} else {
+		encoder.initColIDs()
+		encoder.initOffsets()
+	}
+	for i, colID := range encoder.tempColIDs {
+		if encoder.values[i].IsNull() {
+			if r.large {
+				r.colIDs32[nullIdx] = uint32(colID)
+			} else {
+				r.colIDs[nullIdx] = byte(colID)
+			}
+			nullIdx++
+		} else {
+			if r.large {
+				r.colIDs32[notNullIdx] = uint32(colID)
+			} else {
+				r.colIDs[notNullIdx] = byte(colID)
+			}
+			encoder.values[notNullIdx] = encoder.values[i]
+			notNullIdx++
+		}
+	}
+	if r.large {
+		largeNotNullSorter := (*largeNotNullSorter)(encoder)
+		sort.Sort(largeNotNullSorter)
+		if r.numNullCols > 0 {
+			largeNullSorter := (*largeNullSorter)(encoder)
+			sort.Sort(largeNullSorter)
+		}
+	} else {
+		smallNotNullSorter := (*smallNotNullSorter)(encoder)
+		sort.Sort(smallNotNullSorter)
+		if r.numNullCols > 0 {
+			smallNullSorter := (*smallNullSorter)(encoder)
+			sort.Sort(smallNullSorter)
+		}
+	}
+	for i := 0; i < notNullIdx; i++ {
+		var err error
+		r.data, err = encodeDatum(r.data, encoder.values[i], encoder.sc)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if len(r.data) > math.MaxUint16 && !r.large {
+			// We need to convert the row to large row.
+			encoder.initColIDs32()
+			for j := 0; j < numCols; j++ {
+				r.colIDs32[j] = uint32(r.colIDs[j])
+			}
+			encoder.initOffsets32()
+			for j := 0; j <= i; j++ {
+				r.offsets32[j] = uint32(r.offsets[j])
+			}
+			r.large = true
+		}
+		if r.large {
+			r.offsets32[i] = uint32(len(r.data))
+		} else {
+			r.offsets[i] = uint16(len(r.data))
+		}
+	}
+	buf = append(buf, CodecVer)
+	flag := byte(0)
+	if r.large {
+		flag = 1
+	}
+	buf = append(buf, flag)
+	buf = append(buf, byte(r.numNotNullCols), byte(r.numNotNullCols>>8))
```

Comment:
Using `binary.LittleEndian.PutUint16` is more clear and safe, which is consistent with `binary.LittleEndian.Uint16` in `setRowData`?

Reply:
We use `append` here, using `binary.LittleEndian.PutUint16` require append two bytes first.